### ISSUE TITLE
NEF - encrypted data blocks

### ISSUE THREAD

**Issue by photopea** [Exiv2/exiv2#621]
Hi, I am trying to understand the NEF format. 

So in the Maker Note, there are some encrypted blocks (e.g. 0x97 = 151 = Color Balance), which is an array of bytes. It starts with a four-letter ASCII like "0205" and at some offset (4 or more, you specify it in "nikonArrayIdx" in makernote_int.cpp) the encrypted data starts (which we can decrypt).

Does this data (after decryption) have any regular structure? [Yo describe it](http://www.exiv2.org/tags-nikon.html) as list of IFDs. In my case, I have these bytes for Color Balance:

    [1, 0, 1, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 2, 26, 1, 0, 1, 0, 1, 123, 3, 3, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 3, 31, 3, 50, 5, 36, 4, 14, 4, 63, 2, 191, 4, 128, 4, 3, 4, 66, 2, 69, 0, 76, 0, 151, 0, 29, 2, 57, 2, 78, 0, 135, 0, 144, 2, 116, 2, 28, 65, 66, 68, 66, 67, 66, 68, 66, 3, 131, 0, 255, 0, 0, 0, 166, 32, 16, 3, 66, 33, 0, 3, 51, 0, 115, 0, 0, ...

I don't see an IFD structure in it :( what these data mean?

**Comment by clanmills** [comment]
The honest answer is "I don't know what it means".  However let me explain a little and together we might make progress.  I attach an image taken with my D6300 in my garden.

When I dump the Nikon3 IFD, I get:
```
731 rmills@rmillsmbp:~/temp/Raw $ exiv2 -pv --grep Exif.Nikon3 DSC_0003.NEF
0x0001 Nikon3       Version                     Undefined   4  48 50 49 49
0x0002 Nikon3       ISOSpeed                    Short       2  0 400
0x0004 Nikon3       Quality                     Ascii       8  RAW    
...
0x0095 Nikon3       NoiseReduction              Ascii       5  OFF 
0x0096 Nikon3       LinearizationTable          Undefined 624  (Binary value suppressed)
0x0099 Nikon3       RawImageCenter              Short       2  3008 2008
...
732 rmills@rmillsmbp:~/temp/Raw $ 
```

That metadata is decrypted as described in http://www.exiv2.org/tags-nikon.html

We can locate the "RAW" data in the image with the command `exiv2 -pR DSC_0003.NEF` as follows:
```
STRUCTURE OF TIFF FILE (II): DSC_0003.NEF
 address |    tag                              |      type |    count |    offset | value
      10 | 0x00fe NewSubfileType               |      LONG |        1 |           | 1
      22 | 0x0100 ImageWidth                   |      LONG |        1 |           | 160
...
     238 | 0x014a SubIFDs                      |      LONG |        3 |       416 | 213504 213624 213852
  STRUCTURE OF TIFF FILE (II): DSC_0003.NEF
   address |    tag                              |      type |    count |    offset | value
    213506 | 0x00fe NewSubfileType               |      LONG |        1 |           | 1
...
  END DSC_0003.NEF
  STRUCTURE OF TIFF FILE (II): DSC_0003.NEF
   address |    tag                              |      type |    count |    offset | value
    213626 | 0x00fe NewSubfileType               |      LONG |        1 |           | 0
...
    213818 | 0x9217 SensingMethod                |     SHORT |        1 |           | 2
  END DSC_0003.NEF
  STRUCTURE OF TIFF FILE (II): DSC_0003.NEF
   address |    tag                              |      type |    count |    offset | value
    213854 | 0x00fe NewSubfileType               |      LONG |        1 |           | 1
    213866 | 0x0103 Compression                  |     SHORT |        1 |           | 6
...
  END DSC_0003.NEF
     250 | 0x0214 ReferenceBlackWhite          |  RATIONAL |        6 |       428 | 0/1 255/1 0/1 255/1 0/1 ...
     262 | 0x02bc XMLPacket                    |      BYTE |     2048 |       476 | <?xpacket begin="..." id="W5M0Mp ...
     274 | 0x8769 ExifTag                      |      LONG |        1 |           | 2544
  STRUCTURE OF TIFF FILE (II): DSC_0003.NEF
   address |    tag                              |      type |    count |    offset | value
      2546 | 0x829a ExposureTime                 |  RATIONAL |        1 |      2948 | 10/800
...
      2702 | 0x927c MakerNote                    | UNDEFINED |   152048 |      3088 | Nikon.....II*.....?.........0211 ...
    STRUCTURE OF TIFF FILE (II): MemIo
     address |    tag                              |      type |    count |    offset | value
          10 | 0x0001 Version                      | UNDEFINED |        4 |           | 0211
          22 | 0x0002 ISOSpeed                     |     SHORT |        2 |           | 0 400
...
    END MemIo
      2714 | 0x9286 UserComment                  | UNDEFINED |       44 |      3028 | ASCII...                         ...
...
  END DSC_0003.NEF
     286 | 0x8825 GPSTag                       |      LONG |        1 |           | 155136
     298 | 0x9003 DateTimeOriginal             |     ASCII |       20 |      2524 | 2016:04:20 08:01:05
     310 | 0x9216 TIFFEPStandardID             |      BYTE |        4 |           | ...
END DSC_0003.NEF
```
The maker note is the IFD of length 152048 and offset 3088.  And that's what appears in Exif.Nikon3.  There are other IFDs being decode and they are:
```
737 rmills@rmillsmbp:~/temp/Raw $ exiv2 -pe DSC_0003.NEF | cut -d. -f 2 | sort | uniq
GPSInfo
Image
MakerNote
Nikon3
NikonAf2
NikonCb2b
NikonFi
NikonIi
NikonLd3
NikonMe
NikonPc
NikonPreview
NikonSi02xx
NikonVr
NikonWt
Photo
SubImage1
SubImage2
SubImage3
738 rmills@rmillsmbp:~/temp/Raw $ 
```

As you have mentioned, there is a 0x0097 tag here:
```
739 rmills@rmillsmbp:~/temp/Raw $ exiv2 -pR DSC_0003.NEF | grep 0097
         538 | 0x0097                              | UNDEFINED |     1188 |     17698 | 0219....K.c..'-...".9r...0...~H[ ...
740 rmills@rmillsmbp:~/temp/Raw $ 
```
1188 bytes of something starting with the ascii bytes 0219, however I don't know what it means.   A binary array of some kind, perhaps?  A lookup table.  I don't know.

[DSC_0003.NEF.zip](https://github.com/Exiv2/exiv2/files/2712640/DSC_0003.NEF.zip)

**Comment by photopea** [comment]
These 1188 bytes at 0x97 are "NikonCb2b" - see here: http://www.exiv2.org/tags-nikon.html  at "Nikon Color Balance 2b Tags". These bytes should contain "Exif.NikonCb2b.WB_RGGBLevels" at "tag 0x91" (not sure what it means). 

I thought the author of the library or the author of the tags-nikon.html could answer me.

**Comment by clanmills** [comment]
You are correct, we decode that as:
```
757 rmills@rmillsmbp:~/temp/Raw $ exiv2 -pa DSC_0003.NEF | grep Cb2b
Exif.NikonCb2b.Version                       Undefined   4  2.19
Exif.NikonCb2b.WB_RGGBLevels                 Short       4  3172 60468 54460 3188
758 rmills@rmillsmbp:~/temp/Raw $ 
```

Andreas wrote the tiff parsing engine in Exiv2 and no longer participates in the project.

In Exiv2 v0.28, we hope to implement a new tiff parser to support BigTiff (64 bit tiff).  In pursuing that goal, we will explore dark holes in our understanding of how the tiff parser discovers and decodes metadata.  I personally won't be undertaking that project, however I will do my best to assist the engineers involved.

Please be aware that most of the open-source community's knowledge of maker notes and raw formats is based on reverse engineering.  For example, Canon recently released a new .CR3 file format for which we believe they have a SDK which they share with Adobe and others.  Sadly, this isn't available to our community.  This is not a criticism of Canon (or any other manufacturer), this is the simply the situation.

I'm going to close this issue, however I'll be happy to continue the discussion if you wish.

**Comment by photopea** [comment]
There has to be a piece of code in Exiv2, which extracts the WB_RGGBLevels out of the binary array. I would like to see and understand how it is done. So there is nobody at Exiv2 now, who understands it?

**Comment by clanmills** [comment]
It's here:
```
nikonmn_int.cpp:        TagInfo( 5, "WB_RGGBLevels", N_("WB RGGB Levels"), N_("WB RGGB levels"), nikonCb2Id, makerTags, unsignedShort, 4, printValue),
nikonmn_int.cpp:        TagInfo( 9, "WB_RGGBLevels", N_("WB RGGB Levels"), N_("WB RGGB levels"), nikonCb2aId, makerTags, unsignedShort, 4, printValue),
nikonmn_int.cpp:  -->   TagInfo(145, "WB_RGGBLevels", N_("WB RGGB Levels"), N_("WB RGGB levels"), nikonCb2bId, makerTags, unsignedShort, 4, printValue),  <---
```
In my file DSC_0003.NEF, the WB_RGGB tag is 145 (0x0091).

```
780 rmills@rmillsmbp:~/temp/Raw $ exiv2  -pv DSC_0003.NEF | grep RG
0x0091 NikonCb2b    WB_RGGBLevels               Short       4  3172 60468 54460 3188
```

When exiv2 encounters tag 0x0091/145, it calls printValue to report the data.  The code called is in src/tags_int.cpp
```
    std::ostream& printValue(std::ostream& os, const Value& value, const ExifData*)
    {
        return os << value;
    }
```

If you execute the following command in the debugger, you'll arrive at printValue()
```
 $ exiv2 -pa --key Exif.NikonCb2b.WB_RGGBLevels /Users/rmills/temp/Raw/DSC_0003.NEF
```

You'll see in the traceback, that you are being called from Exiv2::printList() which is running the vector of Exif metadata and printing the values.

So, may wonder how we created the vector of metadata and that's an interesting question.  It's done by the TiffVisitor.  The tiff parser runs over Tiff Structures.  Exif metadata is encoded as a Tiff.  The tiff parser is recursive because the Makernote is normally another Tiff (embedded in the Exif structure)  As tags are encountered, they are converted to Exiv2::Metadatum and stored in the vector.  I can explain in more detail, if you are interested.

I am trying to help you and it would help to me if I understood why you are so interested in this matter and how deeply you wish to  explore this matter.

**Comment by clanmills** [comment]
Thanks for explaining your use case.  May I suggest you talk to the engineers at darktable as they do raw processing and use Exiv2.  They can probably explain more.

In my file, the cropping appears to be in Exif.NikonAf2 (AutoFocus?)
```
Exif.NikonAf2.Version                        Undefined   4  1.00
Exif.NikonAf2.ContrastDetectAF               Byte        1  On
Exif.NikonAf2.AFAreaMode                     Byte        1  2
Exif.NikonAf2.PhaseDetectAF                  Byte        1  Off
Exif.NikonAf2.PrimaryAFPoint                 Byte        1  0
Exif.NikonAf2.AFPointsUsed                   Byte        7  0 0 0 0 0 0 0
Exif.NikonAf2.AFImageWidth                   Short       1  6000
Exif.NikonAf2.AFImageHeight                  Short       1  4000
Exif.NikonAf2.AFAreaXPosition                Short       1  3000
Exif.NikonAf2.AFAreaYPosition                Short       1  2000
Exif.NikonAf2.AFAreaWidth                    Short       1  984
Exif.NikonAf2.AFAreaHeight                   Short       1  816
```
I think the orientation is in the Exif.Image.Orientation. 
```
833 rmills@rmillsmbp:~/Downloads $ exiv2  -pa ~/temp/Raw/DSC_0003.NEF | grep -i orient
Exif.Image.Orientation                       Short       1  top, left
834 rmills@rmillsmbp:~/Downloads $ 
```

The RowsPerStrip are 4018 and Exif.Nikon3.CropHiSpeed is `0 6016 4016 6016 4016 0 0`.

Turning to your file **portrait.nef**, I see that Width < Length, so presumably it's portrait:
```
844 rmills@rmillsmbp:~/Downloads $ exiv2  -pa --grep width/i --grep length/i --grep height/i ~/Downloads/portrait.nef 
Warning: Directory SubImage1 has an unexpected next pointer; ignored.
Exif.Image.ImageWidth                        Long        1  255
Exif.Image.ImageLength                       Long        1  320
Exif.SubImage1.JPEGInterchangeFormatLength   Long        1  4207463
Exif.SubImage2.ImageWidth                    Long        1  3584
Exif.SubImage2.ImageLength                   Long        1  2844
Exif.Photo.FocalLength                       Rational    1  70.0 mm
Exif.NikonLd2.FocalLength                    Byte        1  71.3 mm
Exif.NikonLd2.MinFocalLength                 Byte        1  24.5 mm
Exif.NikonLd2.MaxFocalLength                 Byte        1  71.3 mm
Exif.NikonFl2.FlashFocalLength               Byte        1  (0)
Exif.NikonAf2.AFImageWidth                   Short       1  0
Exif.NikonAf2.AFImageHeight                  Short       1  0
Exif.NikonAf2.AFAreaWidth                    Short       1  0
Exif.NikonAf2.AFAreaHeight                   Short       1  0
Exif.Photo.FocalLengthIn35mmFilm             Short       1  77.0 mm
845 rmills@rmillsmbp:~/Downloads $ 
```

Curiously, although you have NikonAf2 metadata, there's nothing useful/interesting in your file.  So, how is your file to be cropped?  I loaded the file in MacOSX Preview.app, and it says (in the file info display) General: Pixel Height: 2832, Width: 3552, Orientation: 8 (Rotated 90CW)

When I search the output of Exiv2:
```
857 rmills@rmillsmbp:~/temp/Raw $ exiv2 -pa portrait.nef | grep 28
Warning: Directory SubImage1 has an unexpected next pointer; ignored.
Exif.SubImage2.ImageLength                   Long        1  2844
Exif.SubImage2.RowsPerStrip                  Long        1  2844
Exif.Nikon3.CropHiSpeed                      Short       7  3 4288 2844 3584 2844 352 0
Exif.NikonSi02xx.VibrationReduction          Byte        1  (128)
Exif.Nikon3.LinearizationTable               Undefined  46  70 48 0 8 0 8 0 8 0 8 34 0 183 42 170 157 224 114 94 145 24 245 28 153 101 130 240 175 191 32 210 210 47 198 193 2 167 134 197 90 33 88 216 166 202 54
Exif.Image.IPTCNAA                           Long       33  5898524 1193614083 540 469958658 134231810 942682162 842150192 3932700 825831691 758723890 808466224 5243420 1935755790 1344302703 1682907182 476867685 285242370 1243654594 1852797793 539906080 1818584143 2013404268 1866864640 1682251890 1952539509 1634627433 1934958700 1850679397 35420524 806093021 809119802 909455418 3487797
858 rmills@rmillsmbp:~/temp/Raw $ 
```
The physical image seems to have 2844 Rows.  How did Apple determine Orientation?  

I ran the file through Exiftool:
```
858 rmills@rmillsmbp:~/temp/Raw $ exiftool portrait.nef | grep 28
File Modification Date/Time     : 2018:12:28 06:43:01+00:00
File Access Date/Time           : 2018:12:28 06:51:24+00:00
File Inode Change Date/Time     : 2018:12:28 06:43:07+00:00
Image Height                    : 2844
Rows Per Strip                  : 2844
Crop Hi Speed                   : 5:4 Crop (4288x2844 cropped to 3584x2844 at pixel 352,0)
Output Image Width              : 2832
Image Size                      : 3584x2844
859 rmills@rmillsmbp:~/temp/Raw $ 
```
Ah, he found the Hi Speed Crop.  You could look in the output of Exiftool for more clues.  

```
859 rmills@rmillsmbp:~/temp/Raw $ exiftool -V5 portrait.nef | grep -i crop
  | | | 13) CropHiSpeed = 3 4288 2844 3584 2844 352 0
  | | | | CropData (SubDirectory) -->
  | | | | | CropLeft = 0
  | | | | | CropTop = 0
  | | | | | CropRight = 6016
  | | | | | CropBottom = 4000
  | | | | | CropOutputWidthInches = 10.0266666666667
  | | | | | CropOutputHeightInches = 6.66666666666667
  | | | | | CropScaledResolution = 300
  | | | | | CropSourceResolution = 600
  | | | | | CropOutputResolution = 300
  | | | | | CropOutputScale = 1
  | | | | | CropOutputWidth = 3008
  | | | | | CropOutputHeight = 2000
  | | | | | CropOutputPixels = 6016000
860 rmills@rmillsmbp:~/temp/Raw $ 
```

**Comment by clanmills** [comment]
As there is no feedback from @photopea, I'm going to re-close this issue.

**Comment by photopea** [comment]
Thanks for showing these logs! But many questions still remain unanswered :(

All NEF files have `Exif.Image.Orientation` set to 1 (so it is not useful at all). 

Checking the Width and Height of the preview is a great way to derermine Landscape / Portrait mode! :) But it does not say, which way the camera was rotated (CW or CCW).

Your last log really does not say anything about the image, as all CropXY values are a lot different that the real size 3584 2844 .

I am very thankful for your help! You don't have to struggle with it anymore, I will try to figure it out myself :)

**Comment by photopea** [comment]
When I process "portrait.nef" through Exiftool, it prints

    Rotation : 270

As I see, this information is not in Shot Info, neither in IPTC data, neither in XMP, neither in Exif 0xbc02 ... Do you have any clue, where did they find it? (there is no discussion going on on their GitHub)

**Comment by clanmills** [comment]
Welcome to the world of metadata.  I have no idea why the camera manufacturers do not put their SDKs into the public domain.  I suspect it all goes back to the 1980s when there was strategic advantage in relationships with Adobe and others.  Today, this behaviour seems rather pointless.

I have a very positive relationship with Phil Harvey and he often comments on issues concerning Exiv2.  I'm hoping to visit Canada in 2020 and to visit him.

The open source community are a small band of cooperative people who work without financial reward.  Corporations are secretive and make billions from cameras and image software.  That's how it is.

**Comment by clanmills** [comment]
Another two thoughts:

1) I don't know why you're writing a NEF viewer.  Perhaps you could engage with the very nice people at darktable.  In software (as in many things in life) **_1 + 1 = 3._**  The community's combined knowledge is greater than the sum of the individual parts.  It's almost certain that you know something that they have not discovered.  And of course, they know stuff that you would like to know. 

2) You could also write to Nikon and ask them if they have an SDK for their cameras.  I assume you have a Nikon Camera, so you are a customer and have already given them money.  They might cooperate, ask you to sign an NDA and give you access to their secret sauce.

**Comment by photopea** [comment]
I made a free web-based photo editor [Photopea.com](https://www.Photopea.com). Since it has an excellent support for PSD (Photoshop) files, more and more people use it as a replacement for Photoshop (about 2 millions of users now). And they also expect more and more Photoshop features to work in Photopea.

I implemented the support for .DNG a year ago, and now, I want to implement .NEF and .CR2. I can not use existing EXIF libraries, because I need all the code to be in Javascript.
